In [20]:
import pandas as pd
data = pd.read_csv('/content/sample_data/Life Expectancy Data.csv')

In [21]:
print(data)

          Country  Year      Status  Life expectancy   Adult Mortality  \
0     Afghanistan  2015  Developing              65.0            263.0   
1     Afghanistan  2014  Developing              59.9            271.0   
2     Afghanistan  2013  Developing              59.9            268.0   
3     Afghanistan  2012  Developing              59.5            272.0   
4     Afghanistan  2011  Developing              59.2            275.0   
...           ...   ...         ...               ...              ...   
2933     Zimbabwe  2004  Developing              44.3            723.0   
2934     Zimbabwe  2003  Developing              44.5            715.0   
2935     Zimbabwe  2002  Developing              44.8             73.0   
2936     Zimbabwe  2001  Developing              45.3            686.0   
2937     Zimbabwe  2000  Developing              46.0            665.0   

      infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   \
0                62     0.01    

In [23]:
df = pd.DataFrame(data)
print(df)
df.columns

          Country  Year      Status  Life expectancy   Adult Mortality  \
0     Afghanistan  2015  Developing              65.0            263.0   
1     Afghanistan  2014  Developing              59.9            271.0   
2     Afghanistan  2013  Developing              59.9            268.0   
3     Afghanistan  2012  Developing              59.5            272.0   
4     Afghanistan  2011  Developing              59.2            275.0   
...           ...   ...         ...               ...              ...   
2933     Zimbabwe  2004  Developing              44.3            723.0   
2934     Zimbabwe  2003  Developing              44.5            715.0   
2935     Zimbabwe  2002  Developing              44.8             73.0   
2936     Zimbabwe  2001  Developing              45.3            686.0   
2937     Zimbabwe  2000  Developing              46.0            665.0   

      infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   \
0                62     0.01    

Index(['Country', 'Year', 'Status', 'Life expectancy ', 'Adult Mortality',
       'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B',
       'Measles ', ' BMI ', 'under-five deaths ', 'Polio', 'Total expenditure',
       'Diphtheria ', ' HIV/AIDS', 'GDP', 'Population',
       ' thinness  1-19 years', ' thinness 5-9 years',
       'Income composition of resources', 'Schooling'],
      dtype='object')

In [25]:
df.columns = df.columns.str.replace(' ', '_')
df.columns =df.columns.str.strip()
df.columns

Index(['Country', 'Year', 'Status', 'Life_expectancy_', 'Adult_Mortality',
       'infant_deaths', 'Alcohol', 'percentage_expenditure', 'Hepatitis_B',
       'Measles_', '_BMI_', 'under-five_deaths_', 'Polio', 'Total_expenditure',
       'Diphtheria_', '_HIV/AIDS', 'GDP', 'Population',
       '_thinness__1-19_years', '_thinness_5-9_years',
       'Income_composition_of_resources', 'Schooling'],
      dtype='object')

In [26]:
num_cols = df.select_dtypes(include=['float64', 'int64']).columns
df[num_cols] = df[num_cols].apply(lambda x: x.fillna(x.mean()))

In [27]:
cat_cols = df.select_dtypes(include=['object']).columns
df[cat_cols] = df[cat_cols].apply(lambda x: x.fillna(x.mode()[0]))
df = df.drop_duplicates()

Question 1: Should a country having a lower life expectancy value (<65) increase its healthcare expenditure in order to improve its average lifespan?

In [32]:
import matplotlib.pyplot as plt
import seaborn as sns

In [39]:
life_expectancy_data = df[df['Life_expectancy_'] < 65]

Question 1: Should a country having a lower life expectancy value (<65) increase its healthcare
expenditure in order to improve its average lifespan

In [40]:
correlation = life_expectancy_data['Total_expenditure'].corr(life_expectancy_data['Life_expectancy_'])
print(f'Correlation between Healthcare Expenditure and Life Expectancy: {correlation}')

Correlation between Healthcare Expenditure and Life Expectancy: -0.15062876854783794


No,  a country with a lower life expectancy value (<65) should not increase its healthcare expenditure in order to improve its average lifespan.  Since there is a negative correlation between healthcare spending and life expectancy, then increasing expenditure will not be beneficial.

Question 2: How does Infant and Adult mortality rates affect life expectancy?

In [45]:
mortality_correlation = ['Life_expectancy_', 'infant_deaths', 'Adult_Mortality']
corr_matrix = df[mortality_correlation].corr()
print(corr_matrix)


                  Life_expectancy_  infant_deaths  Adult_Mortality
Life_expectancy_          1.000000      -0.196535        -0.696359
infant_deaths            -0.196535       1.000000         0.078747
Adult_Mortality          -0.696359       0.078747         1.000000


Since the correlation between life expectancy and infant mortality rate is negative but not close to -1. We can conclude that, There is a weak negative correlation between life expectancy and infant deaths (-0.197), suggesting that higher infant mortality rates are associated with slightly lower life expectancy, but the relationship is not as strong as with adult mortality.
Life expectancy has a strong negative correlation with adult mortality (-0.696), indicating that as adult mortality rates increase, life expectancy decreases significantly.



Question 3: Does Life Expectancy have positive or negative correlation with eating habits, lifestyle,
exercise, smoking, drinking alcohol etc

In [51]:
eatinghabits_columns = ['Life_expectancy_', 'Alcohol', '_BMI_', '_HIV/AIDS', '_thinness__1-19_years']
eatinghabits_data = df[eatinghabits_columns]
correlation_matrix = eatinghabits_data.corr()
print(correlation_matrix)

                       Life_expectancy_   Alcohol     _BMI_  _HIV/AIDS  \
Life_expectancy_               1.000000  0.391598  0.559255  -0.556457   
Alcohol                        0.391598  1.000000  0.318070  -0.048650   
_BMI_                          0.559255  0.318070  1.000000  -0.243548   
_HIV/AIDS                     -0.556457 -0.048650 -0.243548   1.000000   
_thinness__1-19_years         -0.472162 -0.416946 -0.532025   0.203922   

                       _thinness__1-19_years  
Life_expectancy_                   -0.472162  
Alcohol                            -0.416946  
_BMI_                              -0.532025  
_HIV/AIDS                           0.203922  
_thinness__1-19_years               1.000000  


Based on the correlation matrix above, I conclude that Life expectancy has a positive correlation with alcohol consumption (0.392), BMI (0.559), suggesting that moderate alcohol consumption and a healthy BMI are associated with higher life expectancy.
Life expectancy has a strong negative correlation with HIV/AIDS prevalence (-0.556) and thinness in 1-19 year-olds (-0.472), indicating that higher rates of HIV/AIDS and thinness are associated with lower life expectancy.

Questio 4: What is the impact of schooling on the lifespan of humans

In [55]:
school_corr = df['Schooling'].corr(df['Life_expectancy_'])
print(f'Correlation between Schooling and Life Expectancy: {school_corr}')

Correlation between Schooling and Life Expectancy: 0.7150663398620056


The correlation between schooling and life expectancy is 0.715, indicating a strong positive relationship. This means that higher levels of schooling are associated with longer life expectancy. Thus, I conlude that schooling has strong impact on the lifespan of humans.

Question 5: Does Life Expectancy have a positive or negative relationship with drinking alcohol

In [ ]:
alchohol_correlation = df['Alcohol'].corr(df['Life_expectancy_'])
print(f'Correlation between Alcohol consumption and Life Expectancy: {alchohol_correlation}')

-0.1156376770657746


Since there is a negative correlation  between alchohol and life_expectancy, then there is no close relationship between alchohol and life expectancy.

Question 6:Do densely populated countries tend to have lower life expectancy

In [57]:
corr_population = df['Life_expectancy_'].corr(df['Population'])
print(f"Correlation between population and life expectancy: {corr_population}")


Correlation between population and life expectancy: -0.019637701509419587


The correlation between population and life expectancy is -0.0196, indicating a very weak negative relationship. This suggests that densely populated countries do not tend to have significantly lower life expectancy.

Question 7: What is the impact of Immunization coverage on life Expectancy?


In [59]:
corr_hepatitis_b = df['corr_matrix['quant__Life expectancy '].sort_values(ascending=False)'].corr(df['Hepatitis_B'])
corr_polio = df['Life_expectancy_'].corr(df['Polio'])
corr_diphtheria = df['Life_expectancy_'].corr(df['Diphtheria_'])

print(f"Correlation between Hepatitis B immunization and life expectancy: {corr_hepatitis_b}")
print(f"Correlation between Polio immunization and life expectancy: {corr_polio}")
print(f"Correlation between Diphtheria immunization and life expectancy: {corr_diphtheria}")


Correlation between Hepatitis B immunization and life expectancy: 0.20377143740026762
Correlation between Polio immunization and life expectancy: 0.46157377544578987
Correlation between Diphtheria immunization and life expectancy: 0.4754183849366064


In [64]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.tree import DecisionTreeRegressor

With all the correlations between polio, diphtheria, and hepatitis b immunizations and life span being positive, we conclude that there is a very strong relationship between immunization coverage and life expectancy.

Question 8: What are the predicting variables actually affecting life expectancy

In [70]:
features = df.columns.drop('Country')
X = pd.get_dummies(df.drop('Country', axis=1), columns=['Status']).fillna(0)
y = df['Life_expectancy_'].fillna(0)


In [71]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


model = LinearRegression()
model.fit(X_train, y_train)
predictions = model.predict(X_test)

In [72]:
mse = mean_squared_error(y_test, predictions)
print(f'Mean Squared Error: {mse}')

Mean Squared Error: 4.9173551641165155e-17


In [74]:
features = X.columns  # Update features to reflect the columns of X after one-hot encoding
coefficients = pd.DataFrame({'Feature': features, 'Coefficient': model.coef_})
print(coefficients.sort_values(by='Coefficient', ascending=False))

                            Feature   Coefficient
1                  Life_expectancy_  1.000000e+00
3                     infant_deaths  1.442883e-10
0                              Year  9.082807e-11
4                           Alcohol  3.189925e-11
11                Total_expenditure  2.987020e-11
8                             _BMI_  1.251941e-11
13                        _HIV/AIDS  5.754032e-12
2                   Adult_Mortality  2.784435e-12
7                          Measles_  1.447743e-14
14                              GDP  1.682397e-15
20                 Status_Developed  4.179314e-16
19                        Schooling  1.685339e-16
17              _thinness_5-9_years  3.078793e-17
15                       Population -7.239291e-18
16            _thinness__1-19_years -3.756216e-16
21                Status_Developing -4.098299e-16
18  Income_composition_of_resources -1.598004e-15
5            percentage_expenditure -2.149002e-14
6                       Hepatitis_B -1.926808e-12


Question 9: Train and test at least 3 regression models and pick the best performing model to predict life
expectancy

In [76]:
from sklearn.preprocessing import StandardScaler, PolynomialFeatures
from sklearn.linear_model import LinearRegression

In [81]:
X = df.drop(['Country', 'Year', 'Status', 'Life_expectancy_'], axis=1)
y = df['Life_expectancy_']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [82]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


In [83]:
models = {
    'Linear Regression': LinearRegression(),
    'Decision Tree Regression': DecisionTreeRegressor(random_state=42),
    'Polynomial Regression': LinearRegression()
}

In [84]:
best_model = None
best_rmse = float('inf')

In [85]:
for name, model in models.items():
    if name == 'Polynomial Regression':
        poly_features = PolynomialFeatures(degree=2)
        X_train_poly = poly_features.fit_transform(X_train_scaled)
        X_test_poly = poly_features.transform(X_test_scaled)
        model.fit(X_train_poly, y_train)
        y_pred = model.predict(X_test_poly)
    else:
        model.fit(X_train_scaled, y_train)
        y_pred = model.predict(X_test_scaled)

    rmse = mean_squared_error(y_test, y_pred, squared=False)
    print(f"{name} RMSE: {rmse}")

    if rmse < best_rmse:
        best_rmse = rmse
        best_model = model

Linear Regression RMSE: 3.9209398281263734
Decision Tree Regression RMSE: 2.3845984353906395
Polynomial Regression RMSE: 2.9786357753218353


In [86]:
print(f"\nBest performing model: {type(best_model).__name__} with RMSE: {best_rmse}")


Best performing model: DecisionTreeRegressor with RMSE: 2.3845984353906395
